packages/storage/src/internals/amplifyAuthConfigAdapter/generateLocationsFromPaths.ts

```
@@ -0,0 +1,73 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import { PathAccess } from '../types/credentials';
+import { BucketInfo } from '../../providers/s3/types/options';
+import { ENTITY_IDENTITY_URL } from '../utils/constants';
+import { StorageAccess } from '../types/common';
+
+const resolvePermissions = (
+	accessRule: Record<string, string[]>,
+	token: boolean,
+	groups?: string,
+) => {
+	if (!token) {
+		return {
+			permission: accessRule.guest,
+		};
+	}
+	if (groups) {
+		const selectedKey = Object.keys(accessRule).find(
+			access => access.includes(groups) || access.includes('authenticated'),
+		);
+
+		return {
+			permission: selectedKey ? accessRule[selectedKey] : undefined,
+		};
+	}
+
+	return {
+		permission: accessRule.authenticated,
+	};
+};
+
+export const generateLocationsFromPaths = ({
+	buckets,
+	tokens,
+	identityId,
+	userGroup,
+}: {
+	buckets: Record<string, BucketInfo>;
+	tokens: boolean;
+	identityId?: string;
+	userGroup?: string;
+}): PathAccess[] => {
+	const locations: PathAccess[] = [];
+
+	for (const [, bucketInfo] of Object.entries(buckets)) {
+		const { bucketName, paths } = bucketInfo;
+		if (!paths) {
+			// Todo: Verify behavior
+			return locations;
+		}
+		for (const [path, accessRules] of Object.entries(paths)) {
+			if (tokens && identityId && path.includes(ENTITY_IDENTITY_URL)) {
```

Nit: this can be moved to a function similar to `resolvePermissions`

---

This is for fail fast but we can create a util for this for sure.